sklearn/neighbors/dist_metrics.pyx

```diff
@@ -1111,7 +1100,13 @@ cdef class PyFuncDistance(DistanceMetric):
         with gil:
             x1arr = _buffer_to_ndarray(x1, size)
             x2arr = _buffer_to_ndarray(x2, size)
-            return self.func(x1arr, x2arr, **self.kwargs)
+            d = self.func(x1arr, x2arr, **self.kwargs)
+            try:
+                return d
+            except TypeError:
```

Comment:
oh, and @GaelVaroquaux – I had the same thought! I had to run a script like the one @yenchenlin1994 suggested to convince myself that it would catch the exception. I suspect the reason it's caught is because Cython produces code which does the type checking in the same block as the ``return`` statement.